Add initial implementation of a calendar trigger

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add initial implementation of a calendar trigger, that allows more flexible automations based on calendar events, an improvement over the existing "binary sensor" style calendar entity that has many limitations (single active event, issues with overlapping events, etc)

This initial PR only adds support for triggering based on event start, but has the scaffolding in place to allow it to also
work on end time, and support offsets in follow up PRs. There are are additional corner cases to consider for those scenarios, so it seemed worthwhile to first get in all the basic scaffolding and test harnesses so we can do more careful review of how the corner cases are handled on those in more detail.

Related frontend PR: https://github.com/home-assistant/frontend/pull/12343
Related arch discussion: https://github.com/home-assistant/architecture/discussions/700

Overall set of PRs for triggers:
- [X] Typing cleanup (https://github.com/home-assistant/core/pull/68660)
- [X] Add initial trigger for start events (getting scaffolding all setup) (this PR)
- [X] Frontend (https://github.com/home-assistant/frontend/pull/12343)
- [ ] Add trigger for end events
- [ ] Add offset support

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/22328

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

allenporter committed:
Add initial implementation of calendar trigger

This is an initial implementation of a calendar trigger, that supports
triggering on calendar start time.

See architecture proposal in:
https://github.com/home-assistant/architecture/discussions/700

MartinHjelmare commented on homeassistant/components/calendar/trigger.py:
```diff
@@ -0,0 +1,171 @@
+        """Update the set of eligible events."""
+        start_date = now
+        end_date = now + UPDATE_INTERVAL * 2
+        events = await self._entity.async_get_events(self._hass, start_date, end_date)  # type: ignore[no-untyped-call]
```
We shouldn't add type ignores that aren't needed when we'll merge this after the other PR that types this target method.

allenporter replied:
Would you recommend I not send this PR and wait for the other to be resolved next time? Since these were /mostly/ orthogonal i thought it was worth this "hack" that i would cleanup since the other would likely go in fairly quickly. I realize this can cause churn, but took the risk.

MartinHjelmare commented on tests/components/calendar/test_trigger.py:
```diff
@@ -0,0 +1,374 @@
+class FakeSchedule:
+    """Test fixture class for return events in a specific date range."""
+
+    def __init__(self, now: datetime.datetime):
```
```suggestion
    def __init__(self, now: datetime.datetime) -> None:
```

MartinHjelmare commented on tests/components/calendar/test_trigger.py:
```diff
@@ -0,0 +1,374 @@
+    ) -> dict[str, Any]:
+        """Create a new fake event, used by tests."""
+        event_data = {
+            "title": "Event %s" % secrets.token_hex(16),  # Arbitrary unique data
```
Please use f-strings.

allenporter replied:
Is the rationale: Use f-strings whenever the expression is reasonable,  except when it might cause extra performance overhead when not rendered. During tests: always running so use f-strings? I've left debug statements within the code alone for example, but converted the debug statements in the tests to f-strings.

MartinHjelmare replied:
We only use old string format operator for logging statements. Sometimes we use `str.format` in constants. Otherwise we use f-strings.

allenporter commented on homeassistant/components/calendar/trigger.py:
```diff
@@ -0,0 +1,171 @@
+            return pair[0] >= now
+
+        self._events = list(filter(filter_upcoming, event_list))
+        _LOGGER.debug("Populated event list %s", self._events)
```
Fixed.

allenporter commented on homeassistant/components/calendar/trigger.py:
```diff
@@ -0,0 +1,171 @@
+            event_datetime,
+        )
+
+    async def _handle_calendar_event(self, now: Any) -> None:
```
Fixed.

allenporter committed:
Address reviewer feedback

allenporter committed:
Use f-strings for all tests

allenporter committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

decompil3d commented:
Thank you for working on this. It looks like a great start.

MartinHjelmare commented on tests/components/calendar/test_trigger.py:
```diff
@@ -82,7 +82,7 @@ async def async_get_events(
     ) -> list[dict[str, Any]]:
         """Get all events in a specific time frame, used by the demo calendar."""
-        _LOGGER.debug("Fetching events between %s, %s", start_date, end_date)
+        _LOGGER.debug(f"Fetching events between {start_date}, {end_date}")
```
We don't use f-strings in logging statements (due to performance, ie to delay the operation).

allenporter replied:
I've reverted the fstring logging, and then moved this into the main code. I found this pretty essential for debugging  so maybe it will be in production also in the future.

allenporter committed:
Remove logging f-strings, and move to main code

MartinHjelmare commented on tests/components/calendar/test_trigger.py:
```diff
@@ -0,0 +1,373 @@
+
+
+@pytest.fixture(autouse=True)
+async def setup_calendar(hass: HomeAssistant, fake_schedule: FakeSchedule):
```
```suggestion
async def setup_calendar(hass: HomeAssistant, fake_schedule: FakeSchedule) -> None:
```

MartinHjelmare commented:
The architecture discussion needs to be approved before we merge here.

allenporter commented:
> The architecture discussion needs to be approved before we merge here.

Oh, apologies! I would not have sent this PR if I realized as to not waste anyone's time.

I am unclear who the stakeholders are misinterpreted our conversation as consensus. This is not the first time I've done, this so I would appreciate some explicit guidance on what approval looks like Is there more to design work that should happen or specific sets of folks we're looking to get buy in from?

MartinHjelmare commented:
A member should mark the architecture discussion as answered. Normally we want at least two other members to agree besides the discussion author.

Let's wait for Frank's input and then we can go ahead.

allenporter committed:
Remove mypy ignore

balloob commented:
I 👍 the proposal. I think calendar triggers are a great idea.

allenporter committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

allenporter committed:
Update calendar triggers to use new calendar data model

allenporter commented:
This has been updated to use the new calendar data model. I think it was worthwhile to align on the data model first, as it made some of the calls here about what to include in the event payload a little simpler. (basically, all the fields in the event data model)

allenporter commented:
I realize this may also need frontend work too, so i'm going to get started on that now in parallel.

frenck commented:
> A member should mark the architecture discussion as answered. Normally we want at least two other members to agree besides the discussion author.
> 
> Let's wait for Franck's input and then we can go ahead.

I think I missed something here, why was this waiting for me?

Anyways, I'm definitely looking forward to having some things that improve workings with calendars. I'll try to look at this PR and the related architectural discussion this weekend (mobile right now, so not the right environment to do it now).

allenporter commented:
> > A member should mark the architecture discussion as answered. Normally we want at least two other members to agree besides the discussion author.
> > Let's wait for Franck's input and then we can go ahead.
> 
> I think I missed something here, why was this waiting for me?
> 
> Anyways, I'm definitely looking forward to having some things that improve workings with calendars. I'll try to look at this PR and the related architectural discussion this weekend (mobile right now, so not the right environment to do it now).

The context was looking for an additional ACKs on the architecture decision from core members -- I think we have it now, though more input always welcome.

frenck commented on tests/components/calendar/test_trigger.py:
```diff
@@ -0,0 +1,375 @@
+
+TEST_AUTOMATION_ACTION = {
+    "service": "test.automation",
+    "data_template": {
```
```suggestion
    "data": {
```

frenck commented on tests/components/calendar/test_trigger.py:
```diff
@@ -0,0 +1,375 @@
+TEST_UPDATE_INTERVAL = datetime.timedelta(minutes=15)
+
+
+@pytest.fixture
+async def now() -> datetime.datetime:
+    """Fixture to provide a base time for tests."""
+    return dt_util.utcnow()
```
Consider using freezegun instead

allenporter replied:
I am familiar with freezegun for the basic use in a basic test, though having a little trouble knowing the best way to use it for these fixtures.

Right now this test is being a little tricky in that We set a fixed time, then fire fake alarms in the future with wide enough windows that we assume we won't ever hit flakiness while the clock continue to advance. I like the idea of advancing explicitly with freezegun instead, so i agree there is potential here.

For a fixture like `fake_schedule`: should each test freeze its own time and then the fixture calls "now" knowing its getting a frozen time? or just throw a `@pytest.mark.freeze_time` on the `now` fixture, but still advance explicitly during tests with tick? A little more detail, or pointer to an example might help get me moving on it.

allenporter replied:
Figured out something here, done.

frenck commented:
I absolutely love this! (A decent start and much needed part of making calendars more useful)

A bigger /  more useful challenge (I guess) would be a condition (e.g., do I need my car in the next X hours? is the trash picked up tomorrow?)

allenporter commented:
> I absolutely love this! (A decent start and much needed part of making calendars more useful)
> 
> A bigger / more useful challenge (I guess) would be a condition (e.g., do I need my car in the next X hours? is the trash picked up tomorrow?)

Aaaah! I felt like there was a gap here, and now i see what you mean.

As an example, I have a trigger that is based on (sunset AND calendar event active) and wasn't sure how to accomplish this with a trigger only without the binary sensor -- e.g. this blueprint https://community.home-assistant.io/t/evening-light-schedule-driven-by-google-calendar/384550

I'll investigate this

allenporter committed:
Update tests/components/calendar/test_trigger.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

allenporter committed:
Rewrite tests using freezegun

Rewrite tests using freezegun and improve edge case handling, and use utc consistently for all alarms.

MartinHjelmare commented on homeassistant/components/calendar/trigger.py:
```diff
@@ -0,0 +1,172 @@
+        _LOGGER.debug("Calendar event @ %s", now)
+
+        # Consume all events that are eligible to fire
+        while len(self._events) > 0 and self._events[0][0] <= now:
```
```suggestion
        while self._events and self._events[0][0] <= now:
```

MartinHjelmare commented on homeassistant/components/calendar/trigger.py:
```diff
@@ -0,0 +1,172 @@
+    {
+        vol.Required(CONF_PLATFORM): DOMAIN,
+        vol.Required(CONF_ENTITY_ID): cv.entity_id,
+        vol.Required(CONF_EVENT, default=EVENT_START): vol.In({EVENT_START}),
```
```suggestion
        vol.Optional(CONF_EVENT, default=EVENT_START): vol.In({EVENT_START}),
```

allenporter committed:
Update homeassistant/components/calendar/trigger.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

allenporter committed:
Update homeassistant/components/calendar/trigger.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare commented:
Can we cover the non covered cases with tests? Is codecov correct?

allenporter committed:
Increase test coverage based on pr feedback

allenporter commented:
> Can we cover the non covered cases with tests? Is codecov correct?

I've made some tweaks.  It's mostly correct (and a bit more improved now) but let me unpack this:
- Some additional coverage for calendar in general comes from google. I went ahead and added more coverage for the new `to_dict` method here since obviously having coverage in the component is better
- In other cases, there are trade offs for type safety checks or assertions.  e.g. The previous code structure was an attempt to best navigate this discussion https://github.com/home-assistant/developers.home-assistant/pull/1255#discussion_r838520413  --  I interpreted that discussion to air on the side of potentially having less coverage in some cases (e.g. an if + raise would be better than an assertion, possibly resulting in things you can't test).  For example,  `cv.entity_id` adds some guards that the code that calls `component.get_entity` can't know about. In another case, I had some asserts/if guards that maybe that discussion says shouldn't be there.     Curious to see if you think this is an improvement.
- I don't know how to test detaching a trigger and could not see other ex

Current absolute coverage:
```
$ py.test --cov=homeassistant.components.calendar --cov-report=term-missing tests/components/calendar
Name                                            Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------
homeassistant/components/calendar/__init__.py     191     35    82%   54-55, 60-61, 67-69, 123, 130, 166-179, 186-188, 213, 229, 236, 268, 283, 322-323, 325, 329-332, 335
homeassistant/components/calendar/trigger.py       82      4    95%   74-77
-----------------------------------------------------------------------------
TOTAL                                             273     39    86%
```

Current absolute coverage w/ `google` included:
```
$ py.test --cov=homeassistant.components.calendar --cov-report=term-missing tests/components/calendar tests/components/google
Name                                            Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------
homeassistant/components/calendar/__init__.py     191     14    93%   67-69, 174, 187, 213, 229, 236, 322-323, 325, 329-332, 335
homeassistant/components/calendar/trigger.py       82      4    95%   74-77
-----------------------------------------------------------------------------
TOTAL                                             273     18    93%
```

balloob commented:
The easiest way to attach/detach a a trigger is using [`async_initialize_triggers`](https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/trigger.py#L78) from `homeassistant.helpers.trigger` directly. It returns a function to detach it.

MartinHjelmare reviewed (approved):
Ok, thanks!
